AK/Lexer.cpp

```diff
@@ -0,0 +1,332 @@
+    return true;
+}
+
+// Consume a number of characters
+StringView Lexer::consume(size_t count)
+{
+    if (count == 0)
+        return {};
+
+    size_t start = m_index;
+    size_t length = min(count, m_input.length() - m_index);
+    m_index += length;
+
+    return m_input.substring_view(start, length);
+}
+
+// Consume the rest of the input
+StringView Lexer::consume_all()
+{
+    if (eof())
+        return {};
+
+    auto rest = m_input.substring_view(m_index, m_input.length() - m_index);
+    m_index = m_input.length();
+    return rest;
+}
+
+// Consume until a new line is found
+StringView Lexer::consume_line()
+{
+    size_t start = m_index;
+    while (!eof() && peek() != '\r' && peek() != '\n')
+        m_index++;
+    size_t length = m_index - start;
+
+    consume_if_next_is('\r');
+    consume_if_next_is('\n');
+
+    if (length == 0)
+        return {};
+    return m_input.substring_view(start, length);
+}
+
+// Consume and return characters while `condition` returns true
+StringView Lexer::consume_while(Condition condition)
+{
+    size_t start = m_index;
+    while (!eof() && condition(peek()))
+        m_index++;
+    size_t length = m_index - start;
+
+    if (length == 0)
+        return {};
+    return m_input.substring_view(start, length);
+}
+
+// Consume and return characters until `stop` is peek'd
+// The `stop` character is ignored, as it is user-defined
+StringView Lexer::consume_until(char stop)
+{
+    size_t start = m_index;
+    while (!eof() && peek() != stop)
+        m_index++;
+    size_t length = m_index - start;
+
+    ignore();
+
+    if (length == 0)
+        return {};
+    return m_input.substring_view(start, length);
+}
+
+// Consume and return characters until the string `stop` is found
+// The `stop` string is ignored, as it is user-defined
+StringView Lexer::consume_until(const char* stop)
+{
+    size_t start = m_index;
+    while (!eof() && !next_is(stop))
+        m_index++;
+    size_t length = m_index - start;
+
+    ignore(__builtin_strlen(stop));
+
+    if (length == 0)
+        return {};
+    return m_input.substring_view(start, length);
+}
+
+// Consume and return characters until `condition` return true
+StringView Lexer::consume_until(Condition condition)
+{
+    size_t start = m_index;
+    while (!eof() && !condition(peek()))
+        m_index++;
+    size_t length = m_index - start;
+
+    if (length == 0)
+        return {};
+    return m_input.substring_view(start, length);
+}
+
+// Consume a string surrounded by single or double quotes
+// The returned StringView does not include the quotes
+StringView Lexer::consume_quoted_string()
+{
+    ignore_while(is_whitespace);
+
+    if (!is_quote(peek()))
+        return {};
+
+    char quote_char = consume();
+    size_t start = m_index;
+    while (!eof() && peek() != quote_char)
+        m_index++;
+    size_t length = m_index - start;
+
+    ignore();
+
+    return m_input.substring_view(start, length);
```

Comment:
Indeed, this was not intended to be. I guess not having a terminating quote would be considered an error, thus we should return a null StringView. This will be fixed in the next commit.